### PR TITLE
Wip rgw openssl 7

### PR DESCRIPTION
**Pull request by mdw-at-linuxbox**
So this is similar to "wip-rgw-openssl-4" except using civetweb 1.8.

f5f8c99 use civetweb-1.8. (this should be similar to wip-civetweb-1.8...)
5380caf load by soname same as pr#11571.
7da0ed4 document same as pr#11571.
33c1367 now uses "wip-listen1". - note, uses mdw-at-linuxbox fork (+ see https://github.com/ceph/civetweb/pull/15 )
16db4e4 same as last push I did for pr#11571.

### CODE REVIEW

File: doc/install/install-ceph-gateway.rst [hunk 1]
```diff
@@ -145,6 +144,28 @@ execute the following as the ``root`` user::
...
+Before using SSL with civetweb, you will need a certificate that will match
+the host name that that will be used to access the Ceph Object Gateway.
+You may wish to obtain one that has `subject alternate name` fields for
+more flexibility.  If you intend to use S3-style subdomains
+(`Add Wildcard to DNS`_), you will need a `wildcard` certificate.
+
+Civetweb requires that the server key, server certificate, and any other
+CA or intermediate certificates be supplied in one file.  Each of these
+items must be in `pem` form.  Because the combined file contains the
+secret key, it should be protected from unauthorized access.
```
Comment (ktdreyer) [inline doc/install/install-ceph-gateway.rst]: Should we clarify "unauthorized access" here? Explicitly "Only the UID that runs radosgw (the "ceph" UID) should have read access to this key file on the RGW server" ?

Comment (mdw-at-linuxbox) [inline doc/install/install-ceph-gateway.rst]: I'm more worried about people publishing them via the web, making them accessible company wide via puppet, or throwing printouts in the waste basket.  All those "careless" things people do.  So I'd rather not be too specific.

**Comment by mdw-at-linuxbox** [comment]
I've pushed an updated copy of this - based on the civetweb1.8 that's in master.  I've also updated git@github.com:mdw-at-linuxbox/civetweb-1 wip-listen2 aka  https://github.com/ceph/civetweb/pull/15 - that change is required for this.

**Comment by mdw-at-linuxbox** [comment]
This branch appears to be failing due to some unrelated problem with the version of master I based on - causes TEST_rados_get_bad_size_shard_1 to fail in test-erasure-eio.sh, as best I can tell.  (The logic jenkins uses to describe the problem is tagging unrelated "normal" output instead of the problem.)

**Comment by dachary** [comment]
jenkins test this please (no log)

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -870,7 +870,7 @@ endif(WITH_KVS)
 if(WITH_RADOSGW)
   set(civetweb_common_files civetweb/src/civetweb.c)
   add_library(civetweb_common_objs OBJECT ${civetweb_common_files})
-  target_include_directories(civetweb_common_objs PUBLIC
+  target_include_directories(civetweb_common_objs PRIVATE
```
Comment (yehudasa) [inline src/CMakeLists.txt]: @mdw-at-linuxbox why is this needed? Also, please change the title and description of this commit

File: .gitmodules [hunk 3]
```diff
@@ -3,7 +3,8 @@
 	url = https://github.com/ceph/ceph-object-corpus.git
 [submodule "src/civetweb"]
 	path = src/civetweb
-	url = https://github.com/ceph/civetweb
+	url = https://github.com/mdw-at-linuxbox/civetweb
+	branch = wip-listen2
```
Comment (yehudasa) [inline .gitmodules]: @mdw-at-linuxbox please drop this change. Messing with submodule urls is one of the better ways to annoy everyone. We can create a branch on ceph/civetweb instead

File: src/rgw/rgw_rest_s3.cc [hunk 4]
```diff
@@ -3792,10 +3792,12 @@ int RGW_Auth_S3::authorize_v4(RGWRados *store, struct req_state *s)
...
     if (using_qs && (token == "host")) {
-      if (!port.empty() && port != "80" && port != "0") {
-        token_value = token_value + ":" + port;
-      } else if (!secure_port.empty() && secure_port != "443") {
-        token_value = token_value + ":" + secure_port;
+      if (!secure_port.empty()) {
+	if (secure_port != "443")
+	  token_value = token_value + ":" + secure_port;
+      } else if (!port.empty()) {
+	if (secure_port != "80")
```
Comment (yehudasa) [inline src/rgw/rgw_rest_s3.cc]: @mdw-at-linuxbox should it be `port`?

**Comment by mdw-at-linuxbox** [comment]
I've pushed a new version.  This is based on wip-rgw-civetweb-1.8-2.  I may have overwritten somebody else's stuff too - hadn't realized other people were pushing stuff to this branch.  Arg.

So I pushed:
fa69a8f remove useless civetweb include path side effect.
0bbff77 Load libssl.so and libcrypto.so by soname.
f82a98a Document that radosgw now supports SSL.
6e6d7c2 Point to (temporary) civetweb repo with required change.
cbf0533 Handle multiple listening addreses w/ optional ssl "correctly" with civetweb.

Probably I should have corrected the commit message for 6e6d7c2; it's now pointing to
"wip-listen4" in ceph/civetweb not some temporary repo elsewhere.  wip-listen4 is one patch on top of wip-civetweb-1.8-2 for civetweb.

**Comment by mdw-at-linuxbox** [comment]
I've fixed the conflict I had with master, and improved commit wording.

2826aaa rgw: really rebase rgw to civetweb 1.8 now
8308a13 rgw: cmake: remove useless civetweb include path side effect.
7caa0bd0 rgw: civetweb/openssl: Load libssl.so and libcrypto.so by soname.
d4e72df rgw: Document that radosgw now supports SSL.
46ced9d rgw: Get civetweb ssl enhancement: wip-listen4 = wip-civetweb-1.8-2 + mg_get_local_addr
a113cf5 rgw: s3: secure_port should override port, also apply ssl default right.
8bc6dec rgw: Handle multiple listening addreses w/ optional ssl "correctly" with civetweb.

2826aaa - this is the same as the commit in https://github.com/ceph/ceph/pull/12510 - and it will be overridden by the commit in 46ced9d.

a113cf5 - this fixes the code that had been in conflict, basically the logic was not quite right for https.
